torch_frame/nn/encoder/stype_encoder.py

```diff
@@ -685,3 +699,67 @@ def encode_forward(self, feat: TensorData) -> Tensor:
         # -> [batch_size, num_cols, out_channels]
         x = x_lin + self.bias
         return x
+
+
+class Time2VecEncoder(StypeEncoder):
+    r"""Time2VecEncoder for encoding timestamp stype. The input tensor is
+    expected to have 7 columns representing year, month, day, day of week,
+    hour, minute, and second. The encoding includes both a linear component
+    (for the year) and periodic components (for month, day, day of week,
+    hour, minute, second). The linear component applies
+    :class:`torch.nn.Linear` on the year and the periodic components apply
+    sin function on the rest of the columns. And then a
+    :class:`torch.nn.Linear` is applied to the concatenated components.
+    """
+    supported_stypes = {stype.timestamp}
+
+    def __init__(
+        self,
+        out_channels: Optional[int] = None,
+        stats_list: Optional[List[Dict[StatType, Any]]] = None,
+        stype: Optional[stype] = None,
+        post_module: Optional[Module] = None,
+        na_strategy: Optional[NAStrategy] = None,
+    ):
+        super().__init__(out_channels, stats_list, stype, post_module,
+                         na_strategy)
+
+    def init_modules(self):
+        super().init_modules()
+        self.linear_1 = Linear(1, 1)
+        self.periodic = Linear(stype.timestamp.timestamp_num_cols - 1,
+                               stype.timestamp.timestamp_num_cols - 1)
+        self.min_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][0]
+            for i in range(len(self.stats_list))
+        ])
+        self.max_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][1]
+            for i in range(len(self.stats_list))
+        ])
+        self.linear_2 = Linear(stype.timestamp.timestamp_num_cols,
+                               self.out_channels)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        super().reset_parameters()
+        self.linear_1.reset_parameters()
+        self.linear_2.reset_parameters()
+
+    def encode_forward(self, feat: Tensor) -> Tensor:
+        device = feat.device
+        feat[:, :, 0] = (feat[:, :, 0] - self.min_year.to(device)) / (
+            self.max_year.to(device) - self.min_year.to(device) + 1)
+        feat[:, :, 1] -= 1
+        feat[:, :, 1] /= NUM_MONTHS_PER_YEAR
+        feat[:, :, 2] -= 1
+        feat[:, :, 2] /= NUM_DAYS_PER_MONTH
+        feat[:, :, 3] /= NUM_DAYS_PER_WEEK
+        feat[:, :, 4] /= NUM_HOURS_PER_DAY
+        feat[:, :, 5] /= NUM_MINUTES_PER_HOUR
+        feat[:, :, 6] /= NUM_SECONDS_PER_MINUTE
```

Comment:
Would be better to somewhere define which index refers to which date information, e.g.:
```python
TIME_TO_INDEX + {
    'YEAR': 0,
    'WEEK': 1,
}
```